core/src/mender-client.c

```diff
@@ -1108,186 +1132,338 @@ mender_check_device_compatibility(mender_artifact_ctx_t *mender_artifact_ctx) {
+                }
+                NEXT_STATE;
+                /* fallthrough */
+
+            case MENDER_UPDATE_STATE_CLEANUP:
+                if (mender_client_update_module->callbacks[update_state] != NULL) {
+                    ret = mender_client_update_module->callbacks[update_state](update_state, (mender_update_state_data_t)NULL);
+                }
+                NEXT_STATE;
+                break; /* below is the failure path */
+
+            case MENDER_UPDATE_STATE_ROLLBACK:
+                if (!mender_client_update_module->supports_rollback) {
+                    mender_log_warning("Rollback not supported for artifacts of type '%s'", mender_client_update_module->artifact_type);
+                    ret = MENDER_FAIL;
+                } else if (mender_client_update_module->callbacks[update_state] != NULL) {
+                    ret = mender_client_update_module->callbacks[update_state](update_state, (mender_update_state_data_t)NULL);
+                }
+                NEXT_STATE;
+                /* fallthrough */
+
+            case MENDER_UPDATE_STATE_ROLLBACK_REBOOT:
+                /* Save the next state before running the reboot callback (see
+                 * STATE_REBOOT for details). */
+                mender_storage_save_update_state(MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, mender_client_update_module->artifact_type);
+                ret = mender_client_update_module->callbacks[update_state](update_state, (mender_update_state_data_t)NULL);
+
+                if (MENDER_OK == ret) {
+                    /* now we need to get outside of the loop so that a
+                     * potential asynchronous reboot has a chance to kick in
+                     * after a proper cleanup below */
+                    ret = MENDER_DONE;
+                    goto END;
+                }
+                NEXT_STATE;
+                /* fallthrough */
+
+            case MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT:
+                if (mender_client_update_module->callbacks[update_state] != NULL) {
+                    ret = mender_client_update_module->callbacks[update_state](update_state, (mender_update_state_data_t)NULL);
+                }
+                NEXT_STATE;
+                /* fallthrough */
+
+            case MENDER_UPDATE_STATE_FAILURE:
+                mender_client_publish_deployment_status(deployment_id, MENDER_DEPLOYMENT_STATUS_FAILURE);
+                if (mender_client_update_module->callbacks[update_state] != NULL) {
+                    ret = mender_client_update_module->callbacks[update_state](update_state, (mender_update_state_data_t)NULL);
+                }
+                break;
+
+            case MENDER_UPDATE_STATE_END:
+                /* This is only here to cover all possible values of the
+                 * update_state enum, there is nothing to do here, the while
+                 * loop shall stop when we get here. */
+                break;
         }
```

Comment:
Maybe we could have something like `default: assert(false && "Illegal state!!!");`?

Reply:
I believe it's better to discover unhandled states at compilation time.